File: 3rd Sem/NM/python/guassjordan.py
def gauss_jordan(a, b):
    n = len(b)
    
    # Forward elimination: normalize the pivot row and eliminate other rows
    for i in range(n):
        # Normalize the pivot row (make the diagonal element 1)
        a[i] = [x / a[i][i] for x in a[i]]
        b[i] /= a[i][i]
        
        # Eliminate the i-th column in all other rows
        for j in range(n):
            if i != j:
                ratio = a[j][i]
                a[j] = [a[j][k] - ratio * a[i][k] for k in range(n)]
                b[j] -= ratio * b[i]
    
    return b

# Input: number of variables and the augmented matrix (coefficients and constants)
n = int(input("Enter the number of variables: "))
a = [list(map(float, input(f"Enter row {i + 1} coefficients (space-separated): ").split())) for i in range(n)]
b = list(map(float, input("Enter the constants (space-separated): ").split()))

# Call the Gauss-Jordan method and print the solution
solution = gauss_jordan(a, b)
print("Solution:")
for i in range(n):
    print(f"x{i + 1} = {solution[i]}")

File: 3rd Sem/NM/C-program/guassjordan.c
#include <stdio.h>

void gauss_jordan(int n, double a[n][n], double b[n]) {
    for (int i = 0; i < n; i++) {
        double diag = a[i][i];
        for (int k = 0; k < n; k++) {
            a[i][k] /= diag;
        }
        b[i] /= diag;

        for (int j = 0; j < n; j++) {
            if (i != j) {
                double ratio = a[j][i];
                for (int k = 0; k < n; k++) {
                    a[j][k] -= ratio * a[i][k];
                }
                b[j] -= ratio * b[i];
            }
        }
    }
}

int main() {
    int n;
    printf("Enter the number of variables: ");
    scanf("%d", &n);

    double a[n][n], b[n];

    printf("Enter the coefficients row-wise:\n");
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            scanf("%lf", &a[i][j]);
        }
    }

    printf("Enter the constants: ");
    for (int i = 0; i < n; i++) {
        scanf("%lf", &b[i]);
    }

    gauss_jordan(n, a, b);

    printf("Solution:\n");
    for (int i = 0; i < n; i++) {
        printf("x%d = %.6f\n", i + 1, b[i]);
    }

    return 0;
}
